File: Algorithms/108.convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
    if len(nums) == 0 {
        return nil
    }

    mid := len(nums) / 2
    
    left := nums[:mid]
    right := nums[mid+1:]

    node := &TreeNode{nums[mid], sortedArrayToBST(left), sortedArrayToBST(right)}

    return node
}

File: Algorithms/108.convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


struct TreeNode* helper(int* nums, int left, int right) {
    if (left > right) {
        return NULL;
    }

    // 总是选择中间位置左边的数字作为根节点
    int mid = (left + right) / 2;

    struct TreeNode* root = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    root->val = nums[mid];
    root->left = helper(nums, left, mid - 1);
    root->right = helper(nums, mid + 1, right);
    return root;
}

struct TreeNode* sortedArrayToBST(int* nums, int numsSize) {
    //解法1，中序遍历，总是选择中间位置左边的数字作为根节点
    return helper(nums, 0, numsSize - 1);
}